File: Design-Patterns-Using-Kotlin/Factory/Factory1.kt
// Product interface
interface Product {
    fun operation()
}

// Concrete Product classes
class ConcreteProduct1 : Product {
    override fun operation() {
        println("Concrete Product1 Operation!")
    }
}

class ConcreteProduct2 : Product {
    override fun operation() {
        println("Concrete Product2 Operation!")
    }
}

// Creator abstract class
abstract class Factory {
    abstract fun createProduct(): Product
}

// Concrete Creator classes
class ConcreteFactory1 : Factory() {
    override fun createProduct(): Product {
        return ConcreteProduct1()
    }
}

// Concrete Creator classes
class ConcreteFactory2 : Factory() {
    override fun createProduct(): Product {
        return ConcreteProduct2()
    }
}

// Client code
fun main() {
    val factory1: Factory = ConcreteFactory1()
    val product1: Product = factory1.createProduct()
    product1.operation()

    val factory2: Factory = ConcreteFactory2()
    val product2: Product = factory2.createProduct()
    product2.operation()
}

/* 
Concrete Product1 Operation!
Concrete Product2 Operation!
*/

File: Design-Patterns-Using-Cpp/Factory/Factory1.cpp
#include <iostream>

// Product interface
class Product {
public:
    virtual void operation() = 0;
};

// Concrete Product classes
class ConcreteProduct1 : public Product {
public:
    void operation() override {
        std::cout << "Concrete Product1 Operation!" << std::endl;
    }
};

class ConcreteProduct2 : public Product {
public:
    void operation() override {
        std::cout << "Concrete Product2 Operation!" << std::endl;
    }
};

// Creator abstract class
class Factory {
public:
    virtual Product* createProduct() = 0;
};

// Concrete Creator classes
class ConcreteFactory1 : public Factory {
public:
    Product* createProduct() override {
        return new ConcreteProduct1();
    }
};

class ConcreteFactory2 : public Factory {
public:
    Product* createProduct() override {
        return new ConcreteProduct2();
    }
};

// Client code
int main() {
    Factory* factory1 = new ConcreteFactory1();
    Product* product1 = factory1->createProduct();
    product1->operation();

    Factory* factory2 = new ConcreteFactory2();
    Product* product2 = factory2->createProduct();
    product2->operation();

    delete product1;
    delete product2;
    delete factory1;
    delete factory2;

    return 0;
}

/*
Concrete Product1 Operation!
Concrete Product2 Operation!
*/